Add advanced aircon/climate functionality to IRMQTTServer

crankyoldgit opened the pull request:
- Allow Home Assistant (HA) to handle the most basic of settings of a generic aircon.
- Should allow all/most Common A/C (`IRac` class) features to be set via MQTT & HTTP/HTML.
- Reboot URL (/quitquitquit)
- add HA MQTT discovery message URL (/send_discovery)
- Add an "off" mode to A/C Operation mode to keep HA happy.
- Move info to it's own page and add auto-refresh.
- Move common HTML elements to functions.
- Move/document/add admin functions to their own page.
- Fix firmware update page display.
- HTML pages for displaying/setting the Climate/Common AC object.
- Set climate/aircon modes via HTTP Post to the /aircon/set URL.
- Allow integer values in strToDecodeType().
- Recovery of the climate state from the retained MQTT state, if any.
- Add a timer class for milliseconds.
  * Refactor time to human readable string(s).
- Add more stats and information to the Info page.
- General improvements and other cosmetic tweaks.
- Add some comments/documentation on how to use the new APIs.
- Crude HTML menu system.
- Bump version to `v1.0.0-beta`

* **[BUG]** Fix issues caused by having `MQTT_ENABLE` set to false.

* **[WARN]** Need to update `PubSubClient.h` to have the following (or larger) value:
`#define MQTT_MAX_PACKET_SIZE 768` otherwise HA MQTT climate discovery will not work. (too big)

Tested on my personal IR device. Seems to work as expected. I'm sure there will be bugs. :)

[TODO] Split this up and add some unit test coverage. It's too big now!

crankyoldgit committed:
Add parsing functions for Protocols and common A/C settings.
- Add an "off" mode to A/C Operation mode to keep Home Assistant happy.

crankyoldgit committed:
Prototype Home Assistant Climate functionality to IRMQTTServer
- Progress check-in of code.
- Allows Home Assistant to handle the most basic of settings of a generic aircon.
- Should allow all Common A/C (IRac class) features to be set via MQTT.
- add Reboot URL (/quitquitquit)
- add HA MQTT automatic discovery message URL (/send_discovery)

* Appears to sort of work in limited testing thus far.

crankyoldgit committed:
Progress check-in for IRMQTTServer code.
* Refactor a lot of the MQTT climate defines.
* Refactor the HTML page(s).
  - Move info to it's own page and add auto-refresh.
  - Move common HTML elements to functions.
  - Move/document/add admin functions to their own page.
* Fix firmware update page display.
* Minor tweaks/fixes to the MQTT Climate interface.
* General code cleanup.

crankyoldgit committed:
Add Common A/C page to IRMQTTServer.
Progress check-in.
* HTML pages for displaying/setting the Climate/Common AC object.
TODO: Still needs the form back-end for accepting changes via HTML.

Minor cleanup of code formating.

crankyoldgit committed:
Add processing of Aircon HTML form data.
* Sends codes via HTTP Post to the /aircon/set URL.
* Allow integer values in strToDecodeType().

crankyoldgit committed:
Fix issues caused by having `MQTT_ENABLE` set to false.

crankyoldgit committed:
Add climate state recovery to IRMQTTServer
* Recovery of the climate state from the retained MQTT state, if any.
* Add a timer class for milliseconds.
  - Refactor time to human readable string(s).
* Add more stats and information to the Info page.
  - Climate info
  - MQTT pub/recv message counts.
  - Improved formating.
* Refactor the state broadcast.
  - Ensure broadcasts can't happen while recovering the state.
* General bug fixes/improvements.

crankyoldgit committed:
Add comments & documentation.
* Add some comments/documentation on how to use the new APIs.
* Other cosmetic tweaks.
* Bump it to beta status.

crankyoldgit commented:
FYI @sheppy99

sheppy99 commented:
Thanks. Can’t do any testing for the next month, but things are working with the current version albeit it loses connection on average a couple of times per day. Fortunately the connection losses don’t seem to coincide with commands

crankyoldgit committed:
Change a #define to be more consistent.

crankyoldgit committed:
Fix output of the timeSince() function.

crankyoldgit committed:
Add a build environment for no MQTT
Add a PlatformIO environment that should be built without MQTT so
we know everything works when MQTT_ENABLE is set to false.

crankyoldgit committed:
Add Built/Compiled time to Info page.
